### PR TITLE
HPCC-23079 A blank header/footer should not be treated as default

### PR DESCRIPTION
**Pull request by jakesmith**
If XML('', '') was specified Thor treated that as no
header/footer and defaulted to a default header and footer.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by jakesmith** [comment]
@AttilaVamos - I believe you have a branch/JIRA for new regression tests to test this and other edge cases relating to footer/headers ? Can you link to this JIRA?

@AttilaVamos - please review.

**Comment by AttilaVamos** [comment]
HPCC-23075 is linked.

**Review by AttilaVamos — COMMENTED** [review]
@jakesmith Is this fix for Thor only?

**Comment by jakesmith** [comment]
@AttilaVamos - Yes, hthor and roxie did not have this issue.

File: thorlcr/activities/xmlwrite/thxmlwrite.cpp [hunk 1]
```diff
@@ -70,18 +70,11 @@ class CXmlWriteActivityMaster : public CWriteMasterBase
...
-            size32_t footerLength = supplied.length();
-            if (footerLength == 0 )
-                footerLength = supplied.set(DEFAULTXMLFOOTER).newline().length();
-
+            const char *header = helper->getHeader();
+            size32_t headerLen = header ? strlen(header) : strlen(DEFAULTXMLHEADER)+1; // DEFAULTXMLHEADER+'\n' is default output in slave
+            props.setPropInt(FPheaderLength, headerLen);
+            const char *footer = helper->getFooter();
+            size32_t footerLen = footer ? strlen(footer) : strlen(DEFAULTXMLFOOTER)+1; // DEFAULTXMLFOOTER+'\n' is default output in slave
             props.setPropInt(FPfooterLength, footerLength);
```
Comment (AttilaVamos) [inline thorlcr/activities/xmlwrite/thxmlwrite.cpp]: I think this should be 
props.setPropInt(FPfooterLength, footerLen);

**Review by AttilaVamos — CHANGES_REQUESTED** [review]
@jakesmith there is one issue.

File: thorlcr/activities/xmlwrite/thxmlwrite.cpp (hunk 1 above)
Comment (jakesmith) [reply thorlcr/activities/xmlwrite/thxmlwrite.cpp]: you're right, and made me wonder how it compiled..
It's because there are completely unused member variables...
I'll fix and delete the unused members.

**Comment by jakesmith** [comment]
@AttilaVamos - footerLen fixed.

**Review by AttilaVamos — APPROVED** [review]
@jakesmith Looks good.

**Comment by jakesmith** [comment]
@richardkchapman - ready to merge.

**Comment by richardkchapman** [comment]
Is this behaviour new? If not, it's possible people are relying on it, and it may not be an appropriate fix for a point release?

@jakesmith

**Comment by jakesmith** [comment]
The meta fields were introduced in 5.4 (https://track.hpccsystems.com/browse/HPCC-13331).

Published header/footer lengths are used when despraying,
I think the consequence of them being wrong like this (default +values vs 0), are that it will corrupt the xml being desprayed, with the high probability that the desprayed xml will be unreadable.

So, I don't think they can be relying on this behaviour as such, but still it's not been reported (@AttilaVamos hit in testing), and maybe it is more appropriate to move to a minor release ?

@richardkchapman
